packages/foam-vscode/src/services/variable-resolver.ts

```diff
@@ -19,42 +26,40 @@ const knownFoamVariables = new Set([
   'FOAM_DATE_SECONDS_UNIX',
 ]);
 
-export function substituteVariables(
-  text: string,
-  variables: Map<string, string>
-) {
-  variables.forEach((value, variable) => {
-    const regex = new RegExp(
-      // Matches a limited subset of the the TextMate variable syntax:
-      //  ${VARIABLE}  OR   $VARIABLE
-      `\\\${${variable}}|\\$${variable}(\\W|$)`,
-      // The latter is more complicated, since it needs to avoid replacing
-      // longer variable names with the values of variables that are
-      // substrings of the longer ones (e.g. `$FOO` and `$FOOBAR`. If you
-      // replace $FOO first, and aren't careful, you replace the first
-      // characters of `$FOOBAR`)
-      'g' // 'g' => Global replacement (i.e. not just the first instance)
-    );
-    text = text.replace(regex, `${value}$1`);
-  });
+export function substituteVariables(text: string, variables: Variable[]) {
+  // Assumes/requires that the variables are sorted by start position, and non-overlapping
```

Comment:
Sorting them ourselves doesn't feel like a big overhead, could we do it instead of assuming that's the case?

Reply:
I briefly had a `sort` here, then removed it.

They shouldn't ever be out of order. If it were to ever be a problem, I think I'd rather fix the source of the issue rather than quietly hide it here.

I could remove this comment if you felt that it was drawing too much attention to this requirement.

Reply:
Sounds good, I don't feel strongly about this, happy to go with your recommendation and I agree that as this is the expected behavior, we should address out of order issues where they originate.

I have a minor comment here, that I think is what caused me to feel confused about this function.
The two parameters are intrinsically coupled (as the positions that live in the variables refer to the text), I wonder if this coupling should be more explicit.
Or maybe there should actually be a distinction between a `Variable`, which is just a key/value pair, and a `Substitution`, which is a `Variable` with the added metadata "position" that relates to how to apply it to a text.

E.g.
```
export function substituteVariables(text: ResolvedText) {
   ...
}

interface ResolvedText {
  template: string
  substitutions: Substitution[]
}

interface Substitution {
  variable: Variable
  pos: number
  endPos: number
}
```

Just brainstorming here. Maybe even a simpler change/documentation to the API would do the trick.

But my point is that the `Variable` atm feels like is covering two roles (key/value pair, and substitution in the context of a string), I wonder if splitting them would make some things clearer?

Wdyt?

Reply:
@riccardoferretti Makes sense. I've made two changes that clean this up a bunch.

----

1. I removed `extraVariablesToResolve`. It was not needed anymore:
  * `FOAM_TITLE` was now being [resolved separately](https://github.com/foambubble/foam/blob/08d305b9b0c16508ca226a4d82bcf3bbcea40240/packages/foam-vscode/src/services/templates.ts#L257) in `determineNewNoteFilepath`
  * The presence of `FOAM_SELECTED_TEXT` in `extraVariablesToResolve` was being used to determine whether to append `FOAM_SELECTED_TEXT` to the template. But it could equally use the existence of `FOAM_SELECTED_TEXT` in `givenValues`, so I switched to use that.
  
This means that the confusion around "`Variable`s that weren't present in the template" has been removed, since they no longer exist.

----

> The two parameters are intrinsically coupled (as the positions that live in the variables refer to the text), I wonder if this coupling should be more explicit.

2. I moved the handling of the template text and its variable substitution into `TextmateSnippet`.

This means that `variable-resolver.ts` no longer has to know anything about how to find/substitute variables in a template, and `resolveText` can be simpler too.

One downside of this is that there is now non-trivial changes (new methods) in the vendored VSCode code, making it slightly harder to update when you want to refresh the code from the upstream. 🤷

----

I think it's ready for another round of review.

Reply:
> I removed extraVariablesToResolve. It was not needed anymore:

I like that very much 👍

> I moved the handling of the template text and its variable substitution into TextmateSnippet.

The approach makes a lot of sense. Even though `TextmateSnippet` is very much stateful (which I don't love) it is in line with how that class works, so in the end the solution works really well.

> One downside of this is that there is now non-trivial changes (new methods) in the vendored VSCode code, making it slightly harder to update when you want to refresh the code from the upstream. 🤷

Good point, but I agree that it's a good tradeoff.


I don't have any other comment on the code, for me it's good to be merged.
Is there anything I am missing or that you wanted to add here?

Reply:
:shipit: :ship: :shipit: